File: src/loader.ts
import { glob } from 'glob';
import path from 'path';
import { pathToFileURL } from 'url';
import type { LoadedMigration, Migration } from './types.js';

/**
 * Pattern for migration filenames: prefix-name.ts (e.g., 001-initial-setup.ts)
 */
const MIGRATION_PATTERN = /^(\d+)-(.+)\.ts$/;

/**
 * Extract the numeric prefix from a migration filename.
 * Returns null if the filename doesn't match the expected pattern.
 */
export function extractPrefix(filename: string): number | null {
  const match = filename.match(MIGRATION_PATTERN);
  if (!match) return null;
  return parseInt(match[1], 10);
}

/**
 * Extract the migration ID from a filename (filename without extension).
 */
export function extractId(filename: string): string {
  return path.basename(filename, '.ts');
}

/**
 * Discover migration files in the given directory.
 * Returns files sorted by their numeric prefix.
 */
export async function discoverMigrations(migrationsDir: string): Promise<string[]> {
  const pattern = path.join(migrationsDir, '[0-9]*-*.ts');
  const files = await glob(pattern, { absolute: true });

  // Filter to only files matching the expected pattern and sort by prefix
  return files
    .filter(file => {
      const filename = path.basename(file);
      return MIGRATION_PATTERN.test(filename);
    })
    .sort((a, b) => {
      const prefixA = extractPrefix(path.basename(a)) ?? 0;
      const prefixB = extractPrefix(path.basename(b)) ?? 0;
      return prefixA - prefixB;
    });
}

/**
 * Load a single migration file and validate its exports.
 */
export async function loadMigrationFile(filePath: string): Promise<LoadedMigration> {
  const filename = path.basename(filePath);
  const prefix = extractPrefix(filename);

  if (prefix === null) {
    throw new Error(`Invalid migration filename: ${filename}. Expected format: NNN-name.ts`);
  }

  const id = extractId(filename);

  // Use file URL for dynamic import (required for ESM)
  const fileUrl = pathToFileURL(filePath).href;
  const module = await import(fileUrl) as Migration;

  // Validate the module has the required exports
  if (typeof module.up !== 'function') {
    throw new Error(`Migration ${id} must export an 'up' function`);
  }

  return {
    id,
    prefix,
    filePath,
    module,
  };
}

/**
 * Discover and load all migrations from the given directory.
 */
export async function loadMigrations(migrationsDir: string): Promise<LoadedMigration[]> {
  const files = await discoverMigrations(migrationsDir);
  const migrations: LoadedMigration[] = [];

  for (const file of files) {
    const migration = await loadMigrationFile(file);
    migrations.push(migration);
  }

  return migrations;
}

/**
 * List all available migrations without loading their modules.
 * Useful for quick status checks.
 */
export async function listMigrationFiles(migrationsDir: string): Promise<{ id: string; prefix: number; filePath: string }[]> {
  const files = await discoverMigrations(migrationsDir);

  return files.map(file => {
    const filename = path.basename(file);
    const prefix = extractPrefix(filename) ?? 0;
    const id = extractId(filename);
    return { id, prefix, filePath: file };
  });
}

File: src/loader.rs
use anyhow::{Context, Result};
use glob::glob;
use std::path::Path;

use crate::version::is_valid_version;
use crate::Migration;

/// Discover all migrations in the given directory.
/// Migrations must match the pattern XXXXX-name.ext where XXXXX is a 5-char base36 version
pub fn discover_migrations(dir: &Path) -> Result<Vec<Migration>> {
    // Match 5 alphanumeric characters followed by dash
    let pattern = dir.join("[0-9a-z][0-9a-z][0-9a-z][0-9a-z][0-9a-z]-*");
    let pattern_str = pattern
        .to_str()
        .context("Invalid path for migration directory")?;

    let mut migrations: Vec<Migration> = glob(pattern_str)
        .context("Failed to read glob pattern")?
        .filter_map(|entry| entry.ok())
        .filter(|path| path.is_file())
        .filter_map(|path| {
            let filename = path.file_name()?.to_str()?;
            let version = extract_version(filename)?;
            let id = extract_id(filename);
            Some(Migration {
                id,
                version,
                file_path: path,
            })
        })
        .collect();

    // Sort by version string (lexicographic sort works for base36)
    migrations.sort_by(|a, b| a.version.cmp(&b.version));

    Ok(migrations)
}

/// Extract the version from a migration filename.
/// Returns None if the filename doesn't start with a valid 5-char version.
pub fn extract_version(filename: &str) -> Option<String> {
    if filename.len() < 6 {
        return None;
    }
    // Must have dash after 5-char version
    if filename.as_bytes().get(5) != Some(&b'-') {
        return None;
    }
    let version = &filename[..5];
    if is_valid_version(version) {
        Some(version.to_string())
    } else {
        None
    }
}

/// Extract the migration ID from a filename.
/// The ID is the filename without extension (e.g., "1f72f-init" from "1f72f-init.sh")
pub fn extract_id(filename: &str) -> String {
    // Remove extension if present
    match filename.rfind('.') {
        Some(pos) => filename[..pos].to_string(),
        None => filename.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_version() {
        assert_eq!(extract_version("1f72f-init.sh"), Some("1f72f".to_string()));
        assert_eq!(
            extract_version("00000-something.ts"),
            Some("00000".to_string())
        );
        assert_eq!(extract_version("zzzzz-last.py"), Some("zzzzz".to_string()));
        assert_eq!(extract_version("ab-invalid.sh"), None); // Too short
        assert_eq!(extract_version("1234-short.sh"), None); // 4 chars, not 5
        assert_eq!(extract_version("123456-toolong.sh"), None); // No dash at position 5
    }

    #[test]
    fn test_extract_id() {
        assert_eq!(extract_id("1f72f-init.sh"), "1f72f-init");
        assert_eq!(extract_id("00000-add-config.ts"), "00000-add-config");
        assert_eq!(extract_id("zzzzz-no-extension"), "zzzzz-no-extension");
    }
}
